arch/arm/include/cortex_m/tz.h

```diff
@@ -0,0 +1,312 @@
+ */
+u32_t tz_sau_number_of_regions_get(void);
+
+#if defined(CONFIG_ARM_SAU)
+/**
+ *
+ * @brief SAU Region configuration
+ *
+ * A convenient struct to include all required elements
+ * for a SAU region configuration.
+ */
+typedef struct {
+	u8_t region_num;
+	u8_t enable:1;
+	u8_t nsc:1;
+	u32_t base_addr;
+	u32_t limit_addr;
+} tz_sau_conf_t;
+
+
+/**
+ *
+ * @brief Configure SAU Region
+ *
+ * Configure an existing ARMv8-M SAU region.
+ *
+ * @param p_sau_conf pointer to a tz_sau_conf_t structure
+ *
+ * This function shall only be called from Secure state, otherwise the
+ * Non-Secure instances of SAU RNR, RLAR, RBAR registers are RAZ/WI.
+ *
+ * This function shall be called before the Secure Firmware may transition
+ * to Non-Secure state.
+ *
+ * @return 1 if configuration is successful, otherwise 0.
+
+ */
+int tz_sau_region_configure(tz_sau_conf_t *p_sau_conf);
+
+#endif /* CONFIG_ARM_SAU */
+
+/**
+ * @brief Non-Secure function type
+ *
+ * Defines a function pointer type to implement a non-secure function call,
+ * i.e. a function call that switches state from Secure to Non-secure.
+ *
+ * Note:
+ *
+ * A non-secure function call can only happen through function pointers.
+ * This is a consequence of separating secure and non-secure code into
+ * separate executable files.
+ */
+typedef void __attribute__((cmse_nonsecure_call)) (*tz_ns_func_ptr_t) (void);
+
+/* Required for C99 compilation */
+#define typeof  __typeof__
+
+#if defined(CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS)
+/**
+ * @brief Non-Secure entry function attribute.
+ *
+ * Declares a non-secure entry function that may be called from Non-Secure
+ * or from Secure state using the CMSE _cmse_nonsecure_entry intrinsic.
+ *
+ * Note:
+ *
+ * The function must reside in Non-Secure Callable memory region.
+ */
+#define __TZ_NONSECURE_ENTRY_FUNC \
+	__attribute__((cmse_nonsecure_entry, noinline))
+
+#endif /* CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS */
+
+/**
+ * @brief Declare a pointer of non-secure function type
+ *
+ * Note:
+ *
+ * A non-secure function type must only be used as a base type of pointer.
+ */
+#define TZ_NONSECURE_FUNC_PTR_DECLARE(fptr) tz_ns_func_ptr_t fptr
+
+/**
+ * @brief Define a non-secure function pointer
+ *
+ * A non-secure function pointer is a function pointer that has its LSB unset.
+ * The macro uses the CMSE intrinsic: cmse_nsfptr_create(p) to return the
+ * value of a pointer with its LSB cleared.
+ */
+#define TZ_NONSECURE_FUNC_PTR_CREATE(fptr) \
+	((tz_ns_func_ptr_t)(cmse_nsfptr_create(fptr)))
+
+/**
+ * @brief  Check if pointer can be of non-secure function type
+ *
+ * A non-secure function pointer is a function pointer that has its LSB unset.
+ * The macro uses the CMSE intrinsic: cmse_is_nsfptr(p) to evaluate whether
+ * the supplied pointer has its LSB cleared and, thus, can be of non-secure
+ * function type.
+ *
+ * @param fptr supplied pointer to be checked
+ *
+ * @return non-zero if pointer can be of non-secure function type
```

Comment:
Return value of 0 is OK in Zephyr:

```
if (XXX_CHECK()) {
 /* Exception */
  return -ECODE;
}
/* OK */
```

Reply:
Well, that is actually for function operations, where 0 indicates success, negative error codes indicate failures.
This macro is more of a boolean function, i.e. "is <something> <something else>?"

I will rename the define to:
`#define TZ_NONSECURE_FUNC_PTR_IS_NS(fptr) \` to make it more clear.

Reply:
TZ_NONSECURE_FUNC_PTR_IS_NS() looks OK for me.